runtime/mount/subsample.py

```diff
@@ -0,0 +1,432 @@
+            # todo(rteqs): send error to frontend
+            raise ValueError("Too many facets")
+
+    relations: list[DuckDBPyRelation] = []
+
+    categorical_columns = set(
+        conn.query(
+            "select array_agg(column_name) from information_schema.columns where table_name=$table_name and data_type not in ['DOUBLE', 'BIGINT']",
+            params={"table_name": table_name},
+        ).fetchall()[0][0]
+    )
+
+    for i, trace in enumerate(config["traces"]):
+        if trace["type"] != "scattergl" and trace["type"] != "scatter":
+            continue
+
+        x_raw = trace["x"]
+        y_raw = trace["y"]
+
+        x_original = x = quote(trace["x"])
+        y_original = y = quote(trace["y"])
+
+        x_is_categorical = x_raw in categorical_columns
+        y_is_categorical = y_raw in categorical_columns
+
+        color_by = trace.get("color_by")
+        color_by = quote(color_by) if color_by is not None else None
+
+        error_bar = trace.get("error_bar")
+        error_bar = quote(error_bar) if error_bar is not None else None
+
+        marker_size = trace.get("marker_size")
+        marker_size = quote(marker_size) if marker_size is not None else None
+
+        distinct_cols = (
+            f"{x}, {y}"
+            + (f", {facet}" if facet is not None else "")
+            + (f", {color_by}" if color_by is not None else "")
+        )
+
+        cols = (
+            distinct_cols
+            + (f", {min_x} as min_x" if min_x is not None else "")
+            + (f", {max_x} as max_x" if max_x is not None else "")
+            + (f", {min_y} as min_y" if min_y is not None else "")
+            + (f", {max_y} as max_y" if max_y is not None else "")
+            + (f", {error_bar}" if error_bar not in {None, "sem", "stddev"} else "")
+            + (f", {marker_size}" if marker_size is not None else "")
+            + (
+                f", array_value({custom_data_str}) as custom_data"
+                if custom_data_str is not None
+                else ""
+            )
+            + (
+                f", dense_rank() over (order by {x}) as x_dense_rank"
+                if x_is_categorical
+                else ""
+            )
+            + (
+                f", dense_rank() over (order by {y}) as y_dense_rank"
+                if y_is_categorical
+                else ""
+            )
+        )
+
+        x = "x_dense_rank" if x_is_categorical else x
+        y = "y_dense_rank" if y_is_categorical else y
+
+        trace_data = conn.sql(
+            f"""
+            select
+                distinct on ({distinct_cols})
+                {cols}
+            from
+                {table_name}
+            """
+        ).set_alias(f"trace_{i}")
+
+        agg_expr = ""
+        for col, val in [
+            (f"min({x}) as min_x", min_x),
+            (f"max({x}) as max_x", max_x),
+            (f"min({y}) as min_y", min_y),
+            (f"max({y}) as max_y", max_y),
+            # extremas
```

Comment:
```suggestion
            # extrema
```
it's latin (singular would be `extremum`)